src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,322 @@
+            if (!part.isEmpty() || (behavior == QString::KeepEmptyParts))
+                ret += part;
+
+            head = end + sep.size();
+        }
+
+        return ret;
+    }
+
+    QByteArray viewMid(const QByteArray &in, const int pos, const int len = -1)
+    {
+        // mimic QByteArray::mid(pos, len) but instead of returning a full-copy,
+        // we only return a partial view
+
+        if ((pos < 0) || (pos >= in.size()) || (len == 0))
+            return {};
+
+        const int validLen = ((len < 0) || (pos + len) >= in.size())
+            ? in.size() - pos
+            : len;
+        return QByteArray::fromRawData(in.constData() + pos, validLen);
+    }
 }
 
-RequestParser::RequestParser(uint maxContentLength)
-    : m_maxContentLength(maxContentLength)
+RequestParser::RequestParser()
 {
 }
 
-RequestParser::ErrorCode RequestParser::parseHttpRequest(const QByteArray &data, Request &request)
+RequestParser::ParseResult RequestParser::parse(const QByteArray &data)
 {
-    m_request = Request();
+    // Warning! Header names are converted to lowercase
+    return RequestParser().doParse(data);
+}
 
-    // Parse HTTP request header
+RequestParser::ParseResult RequestParser::doParse(const QByteArray &data)
+{
+    // for simplicity, we don't handle malformed requests which use double `LF` as delimiter
     const int headerEnd = data.indexOf(EOH);
     if (headerEnd < 0) {
         qDebug() << Q_FUNC_INFO << "incomplete request";
-        return IncompleteRequest;
+        return {ParseStatus::Incomplete, Request(), 0};
     }
 
-    if (!parseHttpHeader(data.left(headerEnd))) {
+    const QString httpHeaders = QString::fromLatin1(data.constData(), headerEnd);
+    if (!parseStartLines(httpHeaders)) {
         qWarning() << Q_FUNC_INFO << "header parsing error";
-        return BadRequest;
+        return {ParseStatus::BadRequest, Request(), 0};
     }
 
-    // Parse HTTP request message
-    if (m_request.headers.contains("content-length")) {
-        int contentLength = m_request.headers["content-length"].toInt();
-        if (contentLength < 0) {
-            qWarning() << Q_FUNC_INFO << "bad request: content-length is negative";
-            return BadRequest;
+    const int headerLength = headerEnd + EOH.length();
+
+    // handle supported methods
+    if ((m_request.method == HEADER_REQUEST_METHOD_GET) || (m_request.method == HEADER_REQUEST_METHOD_HEAD))
+        return {ParseStatus::OK, m_request, headerLength};
+    if (m_request.method == HEADER_REQUEST_METHOD_POST) {
+        bool ok = false;
+        const int contentLength = m_request.headers[HEADER_CONTENT_LENGTH].toInt(&ok);
+        if (!ok || (contentLength < 0)) {
+            qWarning() << Q_FUNC_INFO << "bad request: content-length invalid";
+            return {ParseStatus::BadRequest, Request(), 0};
         }
-
-        if (contentLength > static_cast<int>(m_maxContentLength)) {
+        if (contentLength > MAX_REQUEST_SIZE) {
```

Comment:
Shouldn't MAX_REQUEST_SIZE be applicable to the entire request, but not just to its content?

Reply:
I'll rename to `MAX_CONTENT_SIZE`.
Under normal usage, limiting content size is the main target here.